### PR TITLE
(NPUP-7, NPUP-12, NPUP-44) Static validation work.

### PR DESCRIPTION
**Pull request by peterhuene**
This is a major refactoring of the native parser to facilitate improved static validation of parsed ASTs.

The bulk of this work is in the two NPUP-44 commits:

* Rewrite the parser rules to make the grammar less permissive of constructs that would be needed to be checked during validation (such as whether or not operands are valid to a binary expression).  This change is the biggest, requiring invasive changes to the AST and parser rules, along with a number of type renames for consistency.
* Refactor out validations that could be done statically into a "validation AST visitor".  Previously the definition scanner and the evaluator were performing checks that could be performed statically during a single pass through the AST, such as parameter validation, check for illegal assignment expressions, check for illegal definition locations, etc.  All of the above is now consolidated into a single validation pass.  Thus, the steps to compile now include a "validate" step (lex -> parse -> validate -> scan -> evaluate).

Additional validations were added (NPUP-12 and NPUP-7).  For NPUP-12, the confusing "is_productive" member of various AST types has been replaced with simply determining if a statement is "effective" (that is, the entire statement has an effect, similar to the `idem` checks in the Ruby implementation).  NPUP-7 added a simple check for illegal assignment expressions in a parameter default value.

### CODE REVIEW

File: lib/tests/fixtures/compiler/parser/bad_capture_param.baseline [hunk 1]
```diff
@@ -0,0 +1,3 @@
+Error: bad_capture_param.pp:1:14: parameter $params is a capture parameter and is not allowed to have a type specifier.
+  function foo(Integer *$params) {
```
Comment (hlindberg) [inline lib/tests/fixtures/compiler/parser/bad_capture_param.baseline]: This is wrong. A capture parameter can be typed. The given type T is taken as Array[T], If it is given as Array[T], then that is the type of the capture. To capture arrays, do Array[Array[T]].

`puppet apply -e 'function foo(Integer *$foo) { notice $foo }; foo(1,2,3)'`

Using Array type to cap the number of entries:
`function foo(Array[Integer, 0, 2] *$foo) { notice $foo }; foo(1,2,3)`
errors as 3 args are given but only 2 accepted.

File: lib/tests/fixtures/compiler/parser/ineffective_function.baseline [hunk 2]
```diff
@@ -0,0 +1,3 @@
+Error: ineffective_function.pp:2:5: ineffective statements may only be used as return values from conditional expressions and functions.
+  1 + 1
```
Comment (hlindberg) [inline lib/tests/fixtures/compiler/parser/ineffective_function.baseline]: This is not quite right. It is allowed to have a .pp file that ends with an ineffctive statement - it is taken as the result of evaluating the content of that file.
This is used for loading of things defined in the puppet language.
In order to trigger the ineffective statement error, you need something like `1+2+3; $a = 10`

File: lib/tests/fixtures/compiler/parser/ineffective.baseline [hunk 3]
```diff
@@ -0,0 +1,3 @@
+Error: ineffective.pp:1:1: ineffective statements may only be used as return values from conditional expressions and functions.
```
Comment (hlindberg) [inline lib/tests/fixtures/compiler/parser/ineffective.baseline]: This is not quite right. It is allowed to have a .pp file that ends with an ineffctive statement - it is taken as the result of evaluating the content of that file.
This is used for loading of things defined in the puppet language.
In order to trigger the ineffective statement error, you need something like `1+2+3; $a = 10`

File: lib/tests/fixtures/compiler/parser/ineffective_in_if.baseline [hunk 4]
```diff
@@ -0,0 +1,3 @@
+Error: ineffective_in_if.pp:1:1: ineffective statements may only be used as return values from conditional expressions and functions.
+  if true {
```
Comment (hlindberg) [inline lib/tests/fixtures/compiler/parser/ineffective_in_if.baseline]: ok at top level in ruby impl since file level is more or less seen as a function

Comment (hlindberg) [inline lib/tests/fixtures/compiler/parser/ineffective_in_if.baseline]: there are several other examples of this below

File: lib/tests/fixtures/compiler/parser/bad_capture_param.baseline (hunk 1 above)
Comment (peterhuene) [inline lib/tests/fixtures/compiler/parser/bad_capture_param.baseline]: Will remove this validation.

File: lib/tests/fixtures/compiler/parser/ineffective.baseline (hunk 3 above)
Comment (peterhuene) [inline lib/tests/fixtures/compiler/parser/ineffective.baseline]: Will remove this restriction and fix up the relevant test cases.

**Pull request merged**